docs/en/configuration.md
# Configuration

> Here we want to explain to you the configuration of the settings of a <mark>LiveProto Client</mark>

---

## Settings

First of all, you need to adjust its settings to your liking

```php
use Tak\Liveproto\Utils\Settings;

$settings = new Settings();


/* Telegram Settings */
$settings->setApiId(5368212);
$settings->setApiHash('87fb8a7a6dc103f87e2e27a30868ce86');
$settings->setDeviceModel('PC 64bit');
$settings->setSystemVersion('4.14.186');
$settings->setAppVersion('1.28.5');
$settings->setSystemLangCode('en-US');
$settings->setLangPack('android');
$settings->setLangCode('en-US');


/* LiveProto Settings ( optional ) */
$settings->setHotReload(false);
$settings->setFloodSleepThreshold(120);
$settings->setReceiveUpdates(false);
$settings->setMaxConnections(10);
$settings->setMinConnections(5);
$settings->setMediaWorkers(3);
$settings->setParallelDownloads(50);
$settings->setParallelUploads(3);
$settings->setAutoCachePeers(false);
$settings->setIPv6(true);
$settings->setTestMode(false);
$settings->setDC(1);
$settings->setProtocol(Tak\Liveproto\Enums\ProtocolType::TcpFull);
$settings->setProxy(type : 'socks5',address : '127.0.0.1:443',username : 'ProxyUser',password : 'ProxyPassword');
$settings->setTakeout(message_users : true,message_chats : true,message_megagroups : true,message_channels : true);
$params = new \Tak\Liveproto\Tl\Types\Other\JsonObject([
	'value'=>array(
		new \Tak\Liveproto\Tl\Types\Other\JsonObjectValue([
			'key'=>'tz_offset',
			'value'=>new \Tak\Liveproto\Tl\Types\Other\JsonNumber([
				'value'=>(float) (new DateTime('now',new DateTimeZone(date_default_timezone_get())))->getOffset()
			])
		])
	)
]);
$settings->setParams($params);
$settings->setSaveTime(3);
$settings->setHideLog(false);
$settings->setMaxSizeLog(100 * 1024 * 1024);
$settings->setPathLog('Liveproto.log');


/* If you want to use MySQL */
$settings->setServer('localhost');
$settings->setUsername('Username');
$settings->setPassword('Password');
$settings->setDatabase('DatabaseName');
```

---

## API ID

- Type : `Integer` <kbd style="color : red">required</kbd>
- Default : `21724`

If this parameter is not set, the api id of an official client is used by default, The API ID you obtained from https://my.telegram.org

!> If your client is not [official](en/configuration.md#Params), using the **API ID** of an [official](en/configuration.md#Params) client is dangerous

## API HASH

- Type : `String` <kbd style="color : red">required</kbd>
- Default : `3e0cb5efcd52300aec5994fdfc5bdc16`

If this parameter is not set, the api hash of an official client is used by default, The API HASH you obtained from https://my.telegram.org

!> If your client is not [official](en/configuration.md#Params), using the **API HASH** of an [official](en/configuration.md#Params) client is dangerous

## Device Model

- Type : `String` <kbd>optional</kbd>
- Default : `$OperatingSystemName`

Defaults to php_uname('s')

## System Version

- Type : `String` <kbd>optional</kbd>
- Default : `$ReleaseName`

Defaults to php_uname('r')

## App Version

- Type : `String` <kbd>optional</kbd>
- Default : `0.26.8.1721-universal`

Default is an [official](en/configuration.md#Params) app version

## System Lang Code

- Type : `String` <kbd>optional</kbd>
- Default : `$Locale`

If the intl extension is enable, it will be used, otherwise the default value is 'en-US'

## Lang Pack

- Type : `String` <kbd>optional</kbd>
- Default : `android`

!> **If you want to use this feature, you must have an [official](en/configuration.md#Params) client !**

## Lang Code

- Type : `String` <kbd>optional</kbd>
- Default : `$Locale`

If the intl extension is enable, it will be used, otherwise the default value is 'en'

## Hot Reload

- Type : `Boolean` <kbd>optional</kbd>
- Default : `True`

If this parameter is set to false, The previous process creates a lock on the session, preventing it from being executed again

?> Note, If false, To be more clear, it executes one process, keeps one in the queue, and cancels subsequent executions OR If true, Every time you run the script, the previous process is disconnected and killed, and a new process is run with the new script and updated code

## Flood Sleep Threshold

- Type : `Integer` <kbd>optional</kbd>
- Default : `120`

If you encounter a FLOOD error, and if it is less than the set value, it will resend the request after X seconds, and if the flood wait time is more than the set time, you will get an error

## Receive Updates

- Type : `Boolean` <kbd>optional</kbd>
- Default : `True`

If this parameter is set to false,use invokeWithoutUpdates(query = initConnection), If you give false value, your session will no longer receive updates

## Max Connections

- Type : `Integer` <kbd>optional</kbd>
- Default : `10`

Specifies the maximum number of connections for the media

!> Increasing the number of connections only helps in cases where the client is trying to upload or download multiple files at the same time

## Min Connections

- Type : `Integer` <kbd>optional</kbd>
- Default : `5`

!> Try not to lower it too much as you may experience problems with upload speed and media upload

Minimum number of connections required for media connections

## Media Workers

- Type : `Integer` <kbd>optional</kbd>
- Default : `3`

!> Higher values may increase download speed but will also increase CPU, memory usage, and the number of simultaneous network connections

Returns the maximum number of parallel worker processes used for downloading / uploading media. This value controls how many concurrent download / upload requests can be executed when fetching / sending large files

## Parallel Downloads

- Type : `Integer` <kbd>optional</kbd>
- Default : `8`

The number of requests sent simultaneously to download the file

!> Increasing it does not always mean increasing download speed and may cause problems

## Parallel Uploads

- Type : `Integer` <kbd>optional</kbd>
- Default : `10`

The number of requests sent simultaneously to upload the file

!> Increasing it may significantly increase resource consumption, but may also increase upload speed

## Auto Cache Peers

- Type : `Boolean` <kbd>optional</kbd>
- Default : `False`

If true, the client initially caches all peers each time so that there is no gap in finding peers

?> I personally recommend enabling this for userbots

!> For bots, the process may take a long time due to the large number of updates. First make sure then try to enable it

## IPv6

- Type : `Boolean` <kbd>optional</kbd>
- Default : `False`

If its value is false, it uses Telegram IP version 4 (ipv4), and if it is true, it uses Telegram IP version 6 (ipv6)

!> Note, You cannot change it later ! For the first time, everything you set is no longer changeable

## Takeout

- Type : `...Mixed` <kbd>optional</kbd>
- Default : `False`

You can pass anything that method [initTakeoutSession](https://tl.liveproto.dev/#/method/account.initTakeoutSession) accepts as a parameter

!> Note, You can also provide an empty array to enable it , like `$settings->setTakeout(array())`

## Test Mode

- Type : `Boolean` <kbd>optional</kbd>
- Default : `False`

If this parameter is set to true, you will be connected to [Telegram's test servers](en/testservers.md)

## DC

- Type : `Integer` <kbd>optional</kbd>
- Default : `0 | Random`

Enter the number of the data center you want to connect to

## Protocol

- Type : `ProtocolType` <kbd>optional</kbd>
- Default : `ProtocolType::FULL`

Set the tcp connection [`protocol`](en/enums.md#ProtocolType)

## Proxy

- Type : `...String` <kbd>optional</kbd>

`SOCKS5` and `HTTP` : You must pass `type` and `address` arguments & `username` and `password` parameters are optional

`SOCKS4` : You must pass `type` and `address` arguments & `user` parameter is optional

`MTPROXY` : You must pass `type` and `address` and `secret` arguments

!> Possible values ​​for the type parameter : `SOCKS5` , `SOCKS4` , `HTTP` , `MTPROXY`

?> **TLS** is set up by adding the letter `s` to the end of each proxy type , like ( case insensitive ) : `SOCKS5s` , `SOCKS4s` , `HTTPs`

```php
$settings->setProxy(type : 'socks4',address : '127.0.0.1:443',user : 'ProxyUser');

$settings->setProxy(url : 'socks5://<username>:<password>@127.0.0.1:443');

$settings->setProxy(type : 'mtproxy',url : 'https://t.me/proxy?server=127.0.0.1&port=443&secret=<secret>');

$settings->setProxy(type : 'https',address : '127.0.0.1:443');

$settings->setProxy(type : 'http',address : '127.0.0.1:80',username : 'ProxyUser',password : 'ProxyPassword');
```

## Params

- Type : `Object | JSONValue` <kbd>optional</kbd>
- Default : `Null`

This parameter is only used for official clients

## Save Time

- Type : `Integer | Float` <kbd>optional</kbd>
- Default : `3`

The value is in seconds and after this time your information is saved in the database

## Hide Log

- Type : `Boolean` <kbd>optional</kbd>
- Default : `False`

If its value is true, then no more logs will be printed

## Max Size Log

- Type : `Integer` <kbd>optional</kbd>
- Default : `10 * 1024 * 1024 | 10MB`

This parameter must be given in bytes, if the size of your log file reaches this value, it will be deleted automatically

## Path Log

- Type : `String` <kbd>optional</kbd>
- Default : `Liveproto.log`

You can set the log file path in which file the logs are saved

### Server

- Type : `Integer` <kbd>optional</kbd>
- Default : `localhost`

The IP of the server on which the database is to be set

### Username

- Type : `String` <kbd style="color : red">required</kbd>

Your database username

### Password

- Type : `String` <kbd style="color : red">required</kbd>

Your database password

### Database

- Type : `String` <kbd>optional</kbd>
- Default : `$Username`

Your database name, By default, the database username value is set for it

docs/CHANGELOG.md
# V0.0.1 PUBLISHED 🎉
 - DATE : 2025-08-03

# V0.0.2 PUBLISHED 🎉
 - DATE : 2025-08-05

# V0.0.3 UPDATED 🎉
 - DATE : 2025-08-09

 - FIX : Interfaces.php

# V0.0.4 UPDATED 🎉
 - DATE : 2025-08-12

 - UPDATE : Layer 211
 - RELEASE : Documentation

# V0.0.5 UPDATED 🎉
 - DATE : 2025-08-17

 - FIX : Bugs
 - RELEASE : Documentation TL

# V0.0.6 UPDATED 🎉
 - DATE : 2025-08-18

# V0.0.7 UPDATED 🎉
 - DATE : 2025-08-19

# V0.0.8 UPDATED 🎉
 - DATE : 2025-08-20

# V0.0.9 UPDATED 🎉
 - DATE : 2025-08-20

# V0.0.10 PUBLISHED 🎉
 - DATE : 2025-09-26

# V0.0.11 PUBLISHED 🎉
 - DATE : 2025-09-01

# V0.0.12 PUBLISHED 🎉
 - DATE : 2025-09-10

# V0.0.12 PUBLISHED 🎉
 - DATE : 2025-09-10

# V0.0.13 PUBLISHED 🎉
 - DATE : 2025-09-12

# V0.0.14 PUBLISHED 🎉
 - DATE : 2025-09-13

# V0.0.15 PUBLISHED 🎉
 - DATE : 2025-09-27

# V0.0.16 PUBLISHED 🎉
 - DATE : 2025-10-16

# V0.0.17 PUBLISHED 🎉
 - DATE : 2025-10-17

# V0.0.18 PUBLISHED 🎉
 - DATE : 2025-10-22

# V0.0.19 PUBLISHED 🎉
 - DATE : 2025-11-07
 
 # V0.0.20 PUBLISHED 🎉
 - DATE : 2025-11-07

# V0.0.21 PUBLISHED 🎉
 - DATE : 2025-11-13

# V0.0.22 PUBLISHED 🎉
 - DATE : 2025-11-13

# V0.0.23 PUBLISHED 🎉
 - DATE : 2025-11-14

# V0.0.24 PUBLISHED 🎉
 - DATE : 2025-12-29

src/Utils/Settings.php
<?php

declare(strict_types = 1);

namespace Tak\Liveproto\Utils;

final class Settings {
	protected array $data;

	public function __get(string $name) : mixed {
		$index = strtolower($name);
		$value = isset($this->data[$index]) ? $this->data[$index] : self::envGuess($index);
		switch($index):
			case 'apiid':
				if(is_int($value) and $value <= 0) throw new \Exception('In the Settings , a valid value for the API ID has not been set');
				is_int($value) || $value = 21724;
				break;
			case 'apihash':
				if(is_string($value) and strlen($value) !== 32) throw new \Exception('In the Settings , a valid value for the API HASH has not been set');
				is_string($value) || $value = '3e0cb5efcd52300aec5994fdfc5bdc16';
				break;
			case 'devicemodel':
				is_string($value) || $value = php_uname('s');
				break;
			case 'systemversion':
				is_string($value) || $value = php_uname('r');
				break;
			case 'appversion':
				is_string($value) || $value = '0.26.8.1721-universal';
				break;
			case 'systemlangcode':
				is_string($value) || $value = (extension_loaded('intl') ? locale_get_primary_language(locale_get_default()).'-'.locale_get_region(locale_get_default()) : 'en-US');
				break;
			case 'langpack':
				is_string($value) || $value = 'android';
				break;
			case 'langcode':
				is_string($value) || $value = (extension_loaded('intl') ? locale_get_primary_language(locale_get_default()) : 'en');
				break;
			case 'hotreload':
				is_bool($value) || $value = true;
				break;
			case 'floodsleepthreshold':
				is_int($value) || $value = 120;
				break;
			case 'receiveupdates':
				is_bool($value) || $value = true;
				break;
			case 'maxconnections':
				is_int($value) || $value = 10;
				$value = min(max($value,1),100);
				break;
			case 'minconnections':
				is_int($value) || $value = 5;
				$value = min(max($value,1),$this->maxConnections);
				break;
			case 'mediaworkers':
				is_int($value) || $value = 3;
				$value = max($value,1);
				break;
			case 'paralleldownloads':
				is_int($value) || $value = 8;
				$value = min(max($value,1),100);
				break;
			case 'paralleluploads':
				is_int($value) || $value = 10;
				$value = min(max($value,1),100);
				break;
			case 'autocachepeers':
				is_bool($value) || $value = false;
				break;
			case 'ipv6':
				is_bool($value) || $value = false;
				break;
			case 'takeout':
				is_array($value) || $value = false;
				break;
			case 'protocol':
				is_a($value,'Tak\\Liveproto\\Enums\\ProtocolType') || $value = null;
				break;
			case 'proxy':
				if(is_array($value) and array_is_list($value) === false):
					if(isset($value['url']) and is_string($value['url'])):
						$parsed = parse_url($value['url']);
						if(isset($parsed['scheme']) and isset($value['type']) === false):
							$value['type'] = $parsed['scheme'];
						endif;
						if(isset($parsed['host'],$parsed['port']) and isset($value['address']) === false):
							$value['address'] = $parsed['host'].chr(58).min($parsed['port'],65535);
						endif;
						if(isset($parsed['user']) and isset($value['username']) === false):
							$value['username'] = $value['user'] = $parsed['user'];
						endif;
						if(isset($parsed['pass']) and isset($value['password']) === false):
							$value['password'] = $parsed['pass'];
						endif;
						if(isset($parsed['query'])):
							parse_str($parsed['query'],$query);
							if(isset($query['server'],$query['port']) and isset($value['address']) === false):
								$value['address'] = $query['server'].chr(58).min(abs(intval($query['port'])),65535);
							endif;
							if(isset($query['secret']) and isset($value['secret']) === false):
								$value['secret'] = $query['secret'];
							endif;
						endif;
					endif;
					if(isset($value['type']) === false) throw new \Exception('The `type` parameter value must be set in the proxy');
					if(is_string($value['type']) === false) throw new \Exception('The value of the `type` parameter in the proxy must be of string type');
					if(isset($value['address']) === false) throw new \Exception('The `address` parameter value must be set in the proxy');
					if(is_string($value['address']) === false) throw new \Exception('The value of the `address` parameter in the proxy must be of string type');
					if(isset($value['username']) === false || is_string($value['username']) === false):
						$value['username'] = null;
					endif;
					if(isset($value['password']) === false || is_string($value['password']) === false):
						$value['password'] = null;
					endif;
					if(isset($value['user']) === false || is_string($value['user']) === false):
						$value['user'] = null;
					endif;
					if(isset($value['secret']) === false || is_string($value['secret']) === false):
						$value['secret'] = null;
					endif;
				else:
					$value = null;
				endif;
				break;
			case 'params':
				is_object($value) || $value = null;
				break;
		endswitch;
		switch($index):
			case 'testmode':
				is_bool($value) || $value = false;
				break;
			case 'dc':
				is_int($value) || $value = 0;
				$value = min(max($value,0),5);
				break;
			case 'savetime':
				is_numeric($value) || $value = 0x3;
				$value = min(max($value,1),60);
				break;
			case 'server':
				is_string($value) || $value = 'localhost';
				break;
			case 'username':
				is_string($value) || $value = strval(null);
				break;
			case 'password':
				is_string($value) || $value = strval(null);
				break;
			case 'database':
				is_string($value) || $value = $this->username;
				break;
		endswitch;
		return $value;
	}
	public function __set(string $name,mixed $value) : void {
		$this->data[strtolower($name)] = $value;
	}
	public function __unset(string $name) : void {
		unset($this->data[strtolower($name)]);
	}
	public function __isset(string $name) : bool {
		return is_null($this->$name) === false;
	}
	public function __call(string $method,array $arguments) : mixed {
		if(preg_match('~^set([a-z0-9]+)$~i',$method,$match)):
			$name = $match[true];
			$this->$name = (array_is_list($arguments) and count($arguments) === 1) ? $arguments[false] : $arguments;
			return $this;
		elseif(preg_match('~^get([a-z0-9]+)$~i',$method,$match)):
			$name = $match[true];
			return $this->$name;
		else:
			throw new \Exception('Call to undefined function '.$method.'()');
		endif;
	}
	public function __debugInfo() : array {
		return $this->data;
	}
	static public function envGuess(string $name,mixed $default = null) : mixed {
		$camel = Tools::snakeTocamel(strtolower($name));
		$snake = Tools::camelTosnake($name);
		$names = array($name,$camel,lcfirst($camel),$snake,strtoupper($snake));
		foreach($names as $name):
			$env = $_ENV[$name] ?? null;
			if(is_null($env) === false):
				return $env;
			endif;
			/* Env can be `true` / `false` / `integer` / `JSON` (array) */
			$env = getenv($name);
			if($env !== false):
				$json = json_decode(strval($env),true);
				return match(true){
					is_numeric($env) => intval($env),
					boolval(is_null($json) === false) => $json,
					default => $env
				};
			endif;
		endforeach;
		return $default;
	}
}

?>